From fd4923ac3c04dfcb0252cee6baea46a777d93b13 Mon Sep 17 00:00:00 2001
From: Andreas Ternstedt <a.ternstedt@setek.se>
Date: Thu, 14 Aug 2025 17:23:43 +0200
Subject: [PATCH 1/4] arm: dts: Add HMX1 device tree for U-Boot

Add device tree for HMX1 board based on imx8mp-var-dart with
HMX-specific hardware configuration:
- Change serial console from UART1 to UART3
- Disable EQOS & FEC ethernet
- Enable USB hubs on boot with proper GPIO configuration
- Add HMX EEPROM support
- Fix SD card detection with correct pinctrl settings

Based on Variscite U-boot version: lf_v2024.04_6.6.52-2.2.0_var01

Co-developed-by: rikardo <rikard.olander@hostmobility.com>
Signed-off-by: Andreas Ternstedt <a.ternstedt@setek.se>

---

Notes:
	v6: Remove EQOS & FEC, no support of marvell 88q2xxx in Uboot
    v5: Adapted patch for Variscite U-boot lf_v2024.04_6.6.52-2.2.0_var01
    v4: - Changed co-cpu binary name
        - Added DDR memory load for co-cpu
        - Delayed M7 firmware loading until after flashing
    v3: - Added product-number to device tree
        - Added RGB LED controller register to device tree
        - Added M7 co-processor support (hmx-m7-fw.bin)
    v2: - Added HMX EEPROM environment variables
        - Added serial/part numbers to device tree
        - Added USB upgrade button support
        - Changed BSP script to hmx_boot.scr
        - Added RGB LED handling
        - Added watchdog initialization
    v1: - Changed UART1 to UART3
        - Updated U-Boot environment for HMX device tree

---
 .../dts/imx8mp-var-dart-dt8mcustomboard.dts   | 103 +++++++++---------
 1 file changed, 54 insertions(+), 49 deletions(-)

diff --git a/arch/arm/dts/imx8mp-var-dart-dt8mcustomboard.dts b/arch/arm/dts/imx8mp-var-dart-dt8mcustomboard.dts
index edb25d64cbb..06b4c46fcd3 100644
--- a/arch/arm/dts/imx8mp-var-dart-dt8mcustomboard.dts
+++ b/arch/arm/dts/imx8mp-var-dart-dt8mcustomboard.dts
@@ -2,20 +2,17 @@
 /*
  * Copyright 2019 NXP
  * Copyright 2020-2025 Variscite Ltd.
+ * Copyright 2022-2024 Host Mobility AB.
+ * Copyright 2025 SETEK Systems AB
  */
 
 #include "imx8mp-var-dart.dtsi"
 
 / {
-	model = "Variscite DART-MX8M-PLUS v1.1 and higher on DT8MCustomBoard 2.0 and higher";
-
-	aliases {
-		gpio5 = &pca6408_1;
-		gpio6 = &pca6408_2;
-	};
+	model = "Host Mobility HMX1";
 
 	chosen {
-		stdout-path = &uart1;
+		stdout-path = &uart3;
 	};
 
 	memory@40000000 {
@@ -38,9 +35,20 @@
 	};
 };
 
-&uart1 {
+&gpio1 {
+	usbhubs_pwr_hog {
+		gpio-hog;
+		gpios = <12 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "usbhubs_pwr_hog";
+		pinctrl-0 = <&pinctrl_usbhubs>;
+		pinctrl-names = "default";
+	};
+};
+
+&uart3 {
 	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_uart1>;
+	pinctrl-0 = <&pinctrl_uart3>;
 	status = "okay";
 };
 
@@ -60,6 +68,12 @@
 		i2c-bus = <0x1>;
 		status = "okay";
 	};
+
+	BR24G04NUX: eeprom@54 {
+		compatible = "atmel,24c04";
+		reg = <0x54>;
+		status = "okay";
+	};
 };
 
 &i2c3 {
@@ -80,20 +94,6 @@
 	scl-gpios = <&gpio5 20 GPIO_ACTIVE_HIGH>;
 	sda-gpios = <&gpio5 21 GPIO_ACTIVE_HIGH>;
 	status = "okay";
-
-	pca6408_1: gpio@20 {
-		compatible = "ti,tca6408";
-		reg = <0x20>;
-		gpio-controller;
-		#gpio-cells = <2>;
-	};
-
-	pca6408_2: gpio@21 {
-		compatible = "ti,tca6408";
-		reg = <0x21>;
-		gpio-controller;
-		#gpio-cells = <2>;
-	};
 };
 
 &usb3_phy0 {
@@ -137,31 +137,7 @@
 };
 
 &eqos {
-	mdio {
-		ethphy1: ethernet-phy@1 {
-			compatible = "ethernet-phy-ieee802.3-c22";
-			reg = <1>;
-			at803x,eee-disabled;
-			eee-broken-1000t;
-			reset-gpios = <&pca6408_2 0 GPIO_ACTIVE_LOW>;
-			reset-assert-us = <10000>;
-			reset-deassert-us = <20000>;
-			vddio-supply = <&vddio1>;
-
-			vddio1: vddio-regulator {
-				regulator-min-microvolt = <1800000>;
-				regulator-max-microvolt = <1800000>;
-			};
-		};
-	};
-};
-
-&fec {
-	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_fec>;
-	phy-mode = "rgmii";
-	phy-handle = <&ethphy1>;
-	status = "okay";
+	status = "disabled";
 };
 
 &usdhc2 {
@@ -176,6 +152,21 @@
 	status = "okay";
 };
 
+#define PE_0_PULL_DISABLE (0<<8)
+#define PE_1_PULL_ENABLE (1<<8)
+#define HYS_0_CMOS (0<<7)
+#define HYS_1_SCHMITT (1<<7)
+#define PUE_0_WEAK_PULL_DOWN (0<<6)
+#define PUE_1_WEAK_PULL_UP (1<<6)
+#define ODE_0_OPEN_DRAIN_DISABLE (0<<5)
+#define ODE_1_OPEN_DRAIN_ENABLE (1<<5)
+#define FSEL_0_SLOW_SLEW_RATE (0<<4)
+#define FSEL_1_FAST_SLEW_RATE (1<<4)
+#define DSE_X1 (0<<1)
+#define DSE_X2 (2<<1)
+#define DSE_X4 (1<<1)
+#define DSE_X6 (3<<1)
+
 &iomuxc {
 	pinctrl_uart1: uart1grp {
 		fsl,pins = <
@@ -184,6 +175,13 @@
 		>;
 	};
 
+	pinctrl_uart3: uart3grp {
+		fsl,pins = <
+			MX8MP_IOMUXC_UART3_RXD__UART3_DCE_RX				0x40
+			MX8MP_IOMUXC_UART3_TXD__UART3_DCE_TX				0x40
+		>;
+	};
+
 	pinctrl_i2c2: i2c2grp {
 		fsl,pins = <
 			MX8MP_IOMUXC_I2C2_SCL__I2C2_SCL					0x400001c2
@@ -226,6 +224,13 @@
 		>;
 	};
 
+	/* 0x146 == pull up, drive strength = 6 */
+	pinctrl_usbhubs: usbhubsgrp {
+		fsl,pins = <
+			MX8MP_IOMUXC_GPIO1_IO12__GPIO1_IO12 				0x146
+		>;
+	};
+
 	pinctrl_reg_usdhc2_vmmc: regusdhc2vmmc {
 		fsl,pins = <
 			MX8MP_IOMUXC_SD2_RESET_B__GPIO2_IO19				0x40
@@ -270,7 +275,7 @@
 
 	pinctrl_usdhc2_gpio: usdhc2gpiogrp {
 		fsl,pins = <
-			MX8MP_IOMUXC_SD2_CD_B__GPIO2_IO12				0x1c4
+			MX8MP_IOMUXC_SD2_CD_B__GPIO2_IO12				(DSE_X2 | HYS_1_SCHMITT | PE_0_PULL_DISABLE)
 		>;
 	};
 
-- 
2.34.1

